pkg/graveler/ref/expired_commits.go

```diff
@@ -0,0 +1,90 @@
+package ref
+
+import (
+	"context"
+	"time"
+
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+type ExpirationDateGetter interface {
+	Get(c *graveler.CommitRecord) time.Time
+}
+
+type ExpiredCommitsFinder struct {
+	branchLister graveler.BranchLister
+	commitGetter graveler.CommitGetter
+	rules        *graveler.RetentionRules
+}
+
+func NewExpiredCommitsFinder(branchLister graveler.BranchLister, commitGetter graveler.CommitGetter, rules *graveler.RetentionRules) *ExpiredCommitsFinder {
+	return &ExpiredCommitsFinder{branchLister: branchLister, commitGetter: commitGetter, rules: rules}
+}
+
+func (e *ExpiredCommitsFinder) GetExpiredCommits(ctx context.Context, repositoryID graveler.RepositoryID, previouslyExpiredCommits []graveler.CommitID) (expired []graveler.CommitID, active []graveler.CommitID, err error) {
+	now := time.Now()
+	processed := make(map[graveler.CommitID]time.Time)
+
+	branchIterator, err := e.branchLister.ListBranches(ctx, repositoryID)
+	if err != nil {
+		return nil, nil, err
+	}
+	previouslyExpiredMap := make(map[graveler.CommitID]bool)
+	for _, commitID := range previouslyExpiredCommits {
+		previouslyExpiredMap[commitID] = true
+	}
+	activeMap := make(map[graveler.CommitID]bool)
+	expiredMap := make(map[graveler.CommitID]bool)
+	for branchIterator.Next() {
+		branchRecord := branchIterator.Value()
+		branchExpirationThreshold := now.AddDate(0, 0, -e.rules.DefaultRetentionDays)
+		if branchExpirationPeriod, ok := e.rules.BranchRetentionDays[branchRecord.BranchID]; ok {
+			branchExpirationThreshold = now.AddDate(0, 0, -branchExpirationPeriod)
+		}
+		commitID := branchRecord.CommitID
+		previousCommit, err := e.commitGetter.GetCommit(ctx, repositoryID, commitID)
+		if err != nil {
+			return nil, nil, err
+		}
+		if previousThreshold, ok := processed[commitID]; ok && !previousThreshold.After(branchExpirationThreshold) {
```

Comment:
Could you replace the 2, I think equality doesn't matter here
```suggestion
		if previousThreshold, ok := processed[commitID]; ok && branchExpirationThreshold.After(previousThreshold) {
```

Reply:
If there is an equality, we have already been to this commit. We don't want to continue the search again from here.